### DISCUSSION TITLE
Importing Piezoelectric events and optogenetic events

### DISCUSSION THREAD

**Discussion by jesusmartin-cortecero** [petersenpeter/CellExplorer#61] [Q&A] [answered]
Hi Peter

I tried to go through the tutorial, but i dont understand very well what do i need to do in order to import stimulation events such as Piezo, Airpuffs, Laser etc, to calculate PSTHs, do I need to enter the "basename.events.mat in the Session_gui to be detected by CellExpllorer? 

Jesus

**Comment by petersenpeter** [comment, marked as answer]
Below steps describes the procedure:

1. Start by defining your events (for P events, dimensions = Px2; in units of seconds): 
```m
stimulation.timestamps(:,1) = [1 5 50 100 1000]';
stimulation.timestamps(:,2) = [1 5 50 100 1000]'+1;
```
Replace the intervals above with your specific stimulation events.

2. Save the events to the basepath (assumed to be current path here)
```m
basename = 'tutorial_session';
saveStruct(stimulation,'events','basename',basename);
```

This will generate an events file in the basepath: tutorial_session.stimulation.events.mat. These two steps should be repeated for each of your types of events (Piezo, Airpuffs, Laser) to generate multiple events file, e.g.
tutorial_session.Piezo.events.mat
tutorial_session.Airpuffs.events.mat
tutorial_session.Laser.events.mat

3. Run ProcessCellMetrics. It will automatically detect events files (tutorial_session.*.events.mat) in the basepath and generate PSTHs
```m
cell_metrics = ProcessCellMetrics;
```
Now, your cell_metrics should contain the PSTH fields: cell_metrics.events.stimulation, and timestamps are saved to cell_metrics.general.events.stimulation

**Reply by jesusmartin-cortecero** [reply]
Hi peter
this is super useful, thanks 
Jesus

**Comment by jesusmartin-cortecero** [comment]
Dear Peter
Sorry for asking  about this question again, somehow i am not able to generate a Psth that looks reasonable to me according to the structure i am recording (VPM with airpuff stimulation)
i wonder if i introduce properly the events in Cell Explorer. 

firsts i define the events as suggested but not sure if i am doing correctly

These time stamps are the stimulation onset in seconds, Right? 
stimulation.timestamps(:,1) = [1 5 50 100 1000]'; 
and these are the offset, right? 
stimulation.timestamps(:,2) = [1 5 50 100 1000]'+1; 

basename = 'Piezo_session';
saveStruct(stimulation,'events','basename',basename);


later when i try to create a customize plot i click in menu spikes/open_spike_data_dialog/add_plot 
but i cannot see the Piezo as an option, then i try selecting events but doesnt look the psth i expected. Times_Vs_SpikeCounts

or sometimes is missaligned

i wonder what i might be doing wrong

Thank you very much in advance

Jesus

**Comment by petersenpeter** [comment]
Hi Jesus

Try to generate the events file and rerun ProcessCellMetrics. This will automatically generate the PSTH
cell_metrics.general.events.* (*=your stimulation name)

If it is not generated, check the ProcessCellMetrics log as it should tell you why it was not done. 

You can also try to generate the PSTH directly, by providing an event struct, and the spikes struct:
PSTH = calc_PSTH(event,spikes);

If you want a PSTH with spike raster, you should define it in the spikes menu. Check this tutorial for help: https://cellexplorer.org/tutorials/plotting-spike-data/#create-and-modify-spike-plots-in-cellexplorer

The two most important things to do is select spikes from the X-data menu, and provide the name of your events:
<img width="782" alt="spikes dialog" src="https://user-images.githubusercontent.com/14089401/193296193-ca95020b-3235-4368-bf06-c0e260fd9e80.png">

**Reply by jesusmartin-cortecero** [reply]
Hi peter. 

Thanks for the tips, i was not providing the Event name in the plot type menu and i was not selecting UID, and now it seems to work, and i get the rasters. 
although, when i import the manipulations data structure i get the following error

"Warning: manipulation timestamps has to be a Nx2 matrix 
 In ProcessCellMetrics (line 332) "

I have the following struct "Tutorial.Piezo.manipulation.mat"
  struct with fields:

    timestamps: [378×2 double]    %1st column is onset timestamps in seconds; 2nd column is offset timestamps
      duration: [378×1 double]       % duration Piezo
             Trial: [378×1 double]        % consecutive number of piezo trials

How should I enter the data to do not get that error?
i tried to import the same data for the structure "events (Tutorial.Piezo.events.mat) however i dont get that error. 

What is the difference between the events structure and the events structure? which one should i import?

Thank you very much in advance

**Reply by petersenpeter** [reply]
Are you sure that the error is not caused by another manipulation struct file? That one look correct to me

**Reply by jesusmartin-cortecero** [reply]
I am not sure, my events structure is the same than the manipulation structure, but only with the manipulation events i get this error. 
i went again through the tutorial and i would like to make sure i understood correctly, when it says PX2 
"timestamps: Px2 matrix with intervals for the P events in seconds." should i enter two columns with the onset and the offset? 

what else could generate that error? 

Another thing problem i have is that the Raster and PSTHs have a misalignment of few milliseconds. i am importing the events from CED Spike2 using the same sampling frequency than it was used for kilosort 30030Hz.
so far i couldnt find out why. 
![image](https://user-images.githubusercontent.com/53047456/193559400-df8c5509-7ecf-443a-9e3a-4aba55f12243.png)
![image](https://user-images.githubusercontent.com/53047456/193552881-1c409d5d-3f42-46ac-9df8-67cfeb6fa0b6.png)
why cluster 40 in the raster plot is in red and the rest of the rasters in blue if all of them were classified as "pyramidal Cell"? 

Sorry for so many questions, i could not find the answers in the Tutorial. 
Thank you very much in advance

**Comment by petersenpeter** [comment]
For all this type of analysis I recommend using rounded sampling rates, so use 30000Hz instead of 30030Hz. 

I don't know what could case the misalignment, the code for importing events and manipulations is the same (except perhaps for alignment, e.g. onset, offset, and center, but that does not seem to be the issue here).

Perhaps you could validate the timestamps with NeuroScope2? Open the session in NeuroScop2, navigate your events/manipulation events, and see if they align properly.
